### PR TITLE
feat: New Feature UpdatePolicy

### PR DESCRIPTION
**Pull request by 00LT00**
Fix: https://github.com/casbin/casbin/issues/584

I am not sure if there is anything missing in my code, please review carefully, thanks

### CODE REVIEW

**Comment by hsluoyz** [comment]
@00LT00 please remove all the "Merge xxx" commits:

image

Then fix the semantic PR:

image

**Comment by 00LT00** [comment]
@hsluoyz Maybe this pr should be merge in master when the major version is updated?

**Comment by hsluoyz** [comment]
@nodece @GopherJ please review.

File: management_api.go [hunk 1]
```diff
@@ -156,6 +159,33 @@ func (e *Enforcer) RemovePolicy(params ...interface{}) (bool, error) {
 	return e.RemoveNamedPolicy("p", params...)
 }
 
+// UpdatePolicy updates an authorization rule from the current policy.
+func(e *Enforcer) UpdatePolicy(p1 string,p2 string) (bool, error){
```
Comment (nodece) [inline management_api.go:163]: p1, p2 should be []string.

File: management_api.go [hunk 2]
```diff
@@ -156,6 +159,33 @@ func (e *Enforcer) RemovePolicy(params ...interface{}) (bool, error) {
...
+	}
+	newPolicy := bytes.Split([]byte(p2),[]byte(","))
+	for k,v := range newPolicy{
+		newPolicy[k] = bytes.TrimSpace(v)
+	}
+
+	return e.UpdatePolicyByName("p",oldPolicy,newPolicy)
+}
+
+func(e *Enforcer) UpdatePolicyByName(ptype string,p1, p2 [][]byte) (bool, error){
```
Comment (nodece) [inline management_api.go:177]: p1, p2 should be []string.

Comment (nodece) [inline management_api.go:177]: p1, p2 should be []string.

File: management_api.go (hunk 1 above)
Comment (00LT00) [reply management_api.go:163]: If use []string, we must use UpdatePolicy by `e.UpdatePolicy([]string{xxx,xxx,xxx}, []string{xxx,xxx,xxx}`.
I think this way is not fine....

File: model/policy.go [hunk 3]
```diff
@@ -156,6 +156,21 @@ func (model Model) RemovePolicy(sec string, ptype string, rule []string) bool {
...
+func (model Model) UpdatePolicy(sec, ptype string, oldRule, newRule []string) bool {
+	oldPolicy := strings.Join(oldRule, DefaultSep)
+	index, ok := model[sec][ptype].PolicyMap[oldPolicy]
+	if !ok {
+		return false
+	}
+
+	model[sec][ptype].Policy[index] = newRule
+	delete(model[sec][ptype].PolicyMap,oldPolicy)
+	model[sec][ptype].PolicyMap[strings.Join(newRule, DefaultSep)] = index
```
Comment (nodece) [inline model/policy.go:169]: I think this is incorrect, you must each the PolicyMap to update the index.

**Comment by nodece** [comment]
@hsluoyz The RP is breaking changes.

File: management_api.go (hunk 1 above)
Comment (nodece) [reply management_api.go:163]: @hsluoyz what do you think?

**Comment by 00LT00** [comment]
ping @hsluoyz

File: management_api.go (hunk 1 above)
Comment (hsluoyz) [reply management_api.go:163]: I think `e.UpdatePolicy([]string{xxx,xxx,xxx}, []string{xxx,xxx,xxx}` is OK for me, if we don't have better solutions.

File: management_api.go (hunk 2 above)
Comment (hsluoyz) [reply management_api.go:177]: Same above.

**Comment by hsluoyz** [comment]
@00LT00 commented.

File: internal_api.go [hunk 4]
```diff
@@ -133,6 +133,43 @@ func (e *Enforcer) removePolicy(sec string, ptype string, rule []string) (bool,
 	return ruleRemoved, nil
 }
 
+func (e *Enforcer) updatePolicy(sec, ptype string, oldRule, newRule []string) (bool, error) {
```
Comment (hsluoyz) [inline internal_api.go:136]: Don't omit []string

File: management_api.go [hunk 5]
```diff
@@ -14,7 +14,9 @@
 
 package casbin
 
-import "github.com/Knetic/govaluate"
+import (
+	"github.com/Knetic/govaluate"
```
Comment (hsluoyz) [inline management_api.go:18]: Don't change it

File: management_api.go [hunk 6]
```diff
@@ -156,6 +158,15 @@ func (e *Enforcer) RemovePolicy(params ...interface{}) (bool, error) {
 	return e.RemoveNamedPolicy("p", params...)
 }
 
+// UpdatePolicy updates an authorization rule from the current policy.
+func (e *Enforcer) UpdatePolicy(p1 []string, p2 []string) (bool, error) {
+	return e.UpdateNamedPolicy("p", p1, p2)
+}
+
+func (e *Enforcer) UpdateNamedPolicy(ptype string, p1, p2 []string) (bool, error) {
```
Comment (hsluoyz) [inline management_api.go:166]: Don't omit []string

File: management_api.go [hunk 7]
```diff
@@ -156,6 +158,15 @@ func (e *Enforcer) RemovePolicy(params ...interface{}) (bool, error) {
 	return e.RemoveNamedPolicy("p", params...)
 }
 
+// UpdatePolicy updates an authorization rule from the current policy.
+func (e *Enforcer) UpdatePolicy(p1 []string, p2 []string) (bool, error) {
```
Comment (hsluoyz) [inline management_api.go:162]: p1 --> rule1

p2 --> rule2

**Comment by hsluoyz** [comment]
@00LT00 reviewed.

File: internal_api.go [hunk 8]
```diff
@@ -133,6 +133,43 @@ func (e *Enforcer) removePolicy(sec string, ptype string, rule []string) (bool,
 	return ruleRemoved, nil
 }
 
+func (e *Enforcer) updatePolicy(sec, ptype string, oldRule []string, newRule []string) (bool, error) {
```
Comment (hsluoyz) [inline internal_api.go:136]: Don't omit, please check all by yourself.

**Comment by hsluoyz** [comment]
> @hsluoyz The RP is breaking changes.

Why is this PR a breaking change?

**Comment by 00LT00** [comment]
ping @nodece

**Comment by nodece** [comment]
@hsluoyz because add `UpdatePolicy` to adapter, so this is breaking changes.

File: persist/adapter.go [hunk 9]
```diff
@@ -56,6 +56,9 @@ type Adapter interface {
 	// RemovePolicy removes a policy rule from the storage.
 	// This is part of the Auto-Save feature.
 	RemovePolicy(sec string, ptype string, rule []string) error
+	// UpdatePolicy updates a policy rule from storage.
+	// This is part of the Auto-Save feature.
+	UpdatePolicy(sec string, ptype string, oldRule, newPolicy []string) error
```
Comment (hsluoyz) [inline persist/adapter.go:61]: Don't add new funcs here. Define a new interface called UpdatableAdapter, add this function to it.

Comment (nodece) [reply persist/adapter.go:61]: @hsluoyz What do you think of merging this PR into beta branch.

Comment (hsluoyz) [reply persist/adapter.go:61]: @nodece that sounds OK to me.

Comment (nodece) [reply persist/adapter.go:61]: @hsluoyz We made a lot of patches, like BatchAdapter. Do you have any good ideas? Maybe we'll make a lot of breaking changes.

Comment (hsluoyz) [reply persist/adapter.go:61]: What're the patches for **BatchAdapter**?

Comment (nodece) [reply persist/adapter.go:61]: @hsluoyz https://github.com/casbin/casbin/blob/master/internal_api.go#L143

Comment (hsluoyz) [reply persist/adapter.go:61]: The patches look ugly but it's OK for me. It works well to support all kinds of adapters and doesn't bring burden or breaking changes to users.

Comment (nodece) [reply persist/adapter.go:61]: > @00LT00 Don't add new funcs here. Define a new interface called UpdatableAdapter, add this function to it.

@00LT00 please add `UpdatableAdapter` to Casbin codebase.

Comment (00LT00) [reply persist/adapter.go]: ok

**Review by hsluoyz — APPROVED** [review]

**Comment by hsluoyz** [comment]
@nodece please review.

**Review by nodece — APPROVED** [review]

File: persist/watcher_update.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package persist
+
+// WatcherEx is the strengthen for Casbin watchers.
```
Comment (nodece) [inline persist/watcher_update.go:17]: @00LT00 please update the comment, this is WatcherUpdatable interface.

File: persist/update_adapter.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package persist
+
+// BatchAdapter is the interface for Casbin adapters with multiple add and remove policy functions.
+type UpdatableAdapter interface {
```
Comment (nodece) [inline persist/update_adapter.go:18]: @00LT00 please update the comment, this is UpdatableAdapter interface.

File: persist/watcher_update.go (hunk 10 above)
Comment (00LT00) [reply persist/watcher_update.go:17]: ok

File: persist/update_adapter.go (hunk 11 above)
Comment (00LT00) [reply persist/update_adapter.go:18]: ok

**Review by nodece — APPROVED** [review]

**Comment by 00LT00** [comment]
@nodece #613  Maybe it can be closed together?

**Comment by nodece** [comment]
@00LT00 #613 has been closed.

**Comment by nodece** [comment]
@00LT00 could you update the https://casbin.org/docs/en/management-api document?

**Pull request merged**